pkg/transforms/mqttsecret.go

```diff
@@ -0,0 +1,236 @@
+	password     string
+	keypemblock  []byte
+	certpemblock []byte
+	capemblock   []byte
+}
+
+const (
+	AuthModeNone             = "none"
+	AuthModeUsernamePassword = "usernamepassword"
+	AuthModeCert             = "cert"
+	AuthModeCA               = "ca"
+)
+
+// NewMQTTSecretSender ...
+func NewMQTTSecretSender(logging logger.LoggingClient, mqttConfig MQTTSecretConfig, persistOnError bool) *MQTTSecretSender {
+	opts := MQTT.NewClientOptions()
+
+	opts.AddBroker(mqttConfig.BrokerAddress)
+	opts.SetClientID(mqttConfig.ClientId)
+	opts.SetAutoReconnect(mqttConfig.AutoReconnect)
+	//avoid casing issues
+	mqttConfig.AuthMode = strings.ToLower(mqttConfig.AuthMode)
+	sender := &MQTTSecretSender{
+		client:         nil,
+		mqttConfig:     mqttConfig,
+		persistOnError: persistOnError,
+		opts:           opts,
+	}
+
+	return sender
+}
+func (sender *MQTTSecretSender) getSecrets(edgexcontext *appcontext.Context) (*mqttSecrets, error) {
+	// No Auth? No Problem!...No secrets required.
+	if sender.mqttConfig.AuthMode == AuthModeNone {
+		return nil, nil
+	}
+
+	secrets, err := edgexcontext.GetSecrets(sender.mqttConfig.SecretPath)
+	if err != nil {
+		return nil, err
+	}
+	mqttSecrets := &mqttSecrets{
+		username:     secrets["user"],
+		password:     secrets["password"],
+		keypemblock:  []byte(secrets["clientkey"]),
+		certpemblock: []byte(secrets["clientcert"]),
+		capemblock:   []byte(secrets["cacert"]),
+	}
+
+	return mqttSecrets, nil
+}
+func (sender *MQTTSecretSender) validateSecrets(secrets mqttSecrets) error {
+	caCertPool := x509.NewCertPool()
+	if sender.mqttConfig.AuthMode == AuthModeUsernamePassword {
+		if secrets.username == "" || secrets.password == "" {
+			return errors.New("AuthModeUsernamePassword selected however username or password was not found at vault path")
+		}
+
+	} else if sender.mqttConfig.AuthMode == AuthModeCert {
+		// need both to make a successful connection
+		if len(secrets.keypemblock) <= 0 || len(secrets.certpemblock) <= 0 {
+			return errors.New("AuthModeCert selected however the key or cert PEM block was not found at vault path")
+		}
+	} else if sender.mqttConfig.AuthMode == AuthModeCA && len(secrets.capemblock) <= 0 {
+		return errors.New("AuthModeCA selected however no PEM Block was found at vault path")
+	}
+
+	if len(secrets.capemblock) > 0 {
+		ok := caCertPool.AppendCertsFromPEM([]byte(secrets.capemblock))
+		if !ok {
+			return errors.New("Error parsing CA Certificate")
+		}
+	}
+	return nil
+}
+func (sender *MQTTSecretSender) configureMQTTClient(secrets mqttSecrets) error {
+	var cert tls.Certificate
+	var err error
+	switch sender.mqttConfig.AuthMode {
+	case AuthModeUsernamePassword:
+		sender.opts.SetUsername(secrets.username)
+		sender.opts.SetPassword(secrets.password)
+	case AuthModeCert:
+		cert, err = tls.X509KeyPair(secrets.certpemblock, secrets.keypemblock)
+		if err != nil {
+			return err
+		}
+		fallthrough // continue on to handle CA
+	case AuthModeCA:
+		caCertPool := x509.NewCertPool()
```

Comment:
Is username/password with CA an option? If so then this doesn't handle that scenario. Maybe move this to an if condition after the switch.

Reply:
restructured it to handle this case.